internal/formats/cyclonedx12xml/encoder.go

```diff
@@ -7,7 +7,7 @@ import (
 	"github.com/anchore/syft/syft/sbom"
 )
 
-func encoder(output io.Writer, s sbom.SBOM) error {
+func encoder(output io.Writer, s sbom.SBOM, _ interface{}) error {
```

Comment:
This is a great suggestion and points out a problem that we have already started to get rid of in the code base... I'll take a closer look at options here 👍

Reply:
TL;DR: I ended up adding an `sbom.Descriptor` that is added to `sbom.SBOM` that contains the application configuration, the tool name, and tool version.

I needed to refresh myself on the input going into the design for format encoding and decoding. The first approach taken (but not used) was to introduce `Encoder`, `Decoder`, and `Validator` interfaces (I re-mocked it up here https://github.com/anchore/syft/commit/5fb0235cfb71fd968af948a3d21b63ed9d972fd1). This approach looked great, however, the drawback was that you couldn't leverage the pro's of this approach. Specifically, you could push format-specific fields (such as application config) down below the interface and let the concrete implementation deal with these difference for each instance. However, that requires that this data is passed at object construction or between construction and the call to `Encode()`, neither of which is ideal since there would be a need to have a list of all formatters to reference... needing to have the data at this time would be difficult.

Function types that are assigned to the Format struct are the next best thing, since it enforces that state must be passed in at encode-time, and not earlier.

I then started to look the other way: since injection of this state via behavior seemed wrong (due to the above notes) I started looking for ways to encapsulate the app config. I had initially thrown out encoding it in `sbom.SBOM` and instead started looking into making a new envelope (something like `format.Context` that would contain `sbom.SBOM` and the application config), however, the more I poked at this the more sane it seemed to put the application config into the SBOM itself.

In syft json, CycloneDX, and SPDX, there are sections that describe the tool that created the SBOM. Now that there are encode and decode workflows we must persist this information in the model or else we will loose it in decoding operations. 

Consider an encode-decode-encode workflow with an old version of syft to a new version of syft... we encode the original SBOM with the old version of syft followed by a decode and encode with the new version of syft. This would be how you convert from an older schema version to a new schema version. Without persisting the original tool name, version, and configuration, we would have no record on how the information was created to begin with. It should only be the `schema` section that changes (to indicate the new format) but not the tool information. 

For these reasons I went the route of adding a new `sbom.Descriptor` section that persists name, version, and optional configuration for the tool that created the SBOM.